---
title: "ITS workflow-Heatmaps"
author: "Sarah Daly"
date: '2024-07-30'
output:
  html_document: default
  word_document: default
editor_options:
  chunk_output_type: console
---

Input:
OTU table (.qza), unfiltered taxonomy table (.qza), rooted photogenic tree (.qza), metadata file (.txt)
Output
.csv and .tif 

All Sequencing Samples Summer Fall Winter Spring

Results focus on Fan/Walls/doors/floors 

#Set Working Directory
```{r}
setwd("C:/Users/dalys/Box/Microbial Food Safety & Spoilage Lab/Lab Members/Sarah Daly/ITS Methods Project/Github upload/Heatmaps")

list.files()
```

#Load Packages
There will be errors/warnings when you load packages but most you can ignore 
```{r}
library(scales)

library(DECIPHER) # This package will help in importing, maintaining, analyzing, manipulating, and exporting a massive amount of sequences.

library(ape) # Analyses of Phylogenetics and Evolution package. Required for tree calculations to be used with phyloseq

library(DESeq2) # This package will help analyze "differential expression" in the microbiota alongside phyloseq

library(ggplot2) # Graphing package used in phyloseq. To edit the default setting of a plot, you need to use functions in this package.

library(phyloseq) # The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.

library(plotly) # A package to create interactive web graphics of use in 3D plots

library(vegan) # The vegan package provides tools for descriptive community ecology. It has most basic functions of diversity analysis, community ordination and dissimilarity analysis. In general, this package is used for Bray-Curtis and Jaccard analyses.

library(tidyverse) # This package is designed to make it easy to install and load multiple 'tidyverse' packages in a single step

library(adespatial) # Tools for the multiscale spatial analysis of multivariate data

library(devtools) # Make package development easier by providing R functions that simplify and expedite common tasks

library(qiime2R) # A package for importing qiime artifacts into an R session

library(microbiome) # Data analysis and visualization

library(microbiomeSeq) # Data analysis and visualization

library("pander") # provide a minimal and easy tool for rendering R objects into Pandoc's markdown

library(ranacapa) # Data analysis 

library(grid) # support data visualization

library(gridExtra)  # support data visualization

library(knitr) # Provides a general-purpose tool for dynamic report generation in R using Literate Programming techniques.

library(png) # Figure download

library("ggdendro") #set of tools for dendrograms and tree plots using 'ggplot2'

library(ggpubr) # publication quality figures, based on ggplot2

library(RColorBrewer) # nice color options

library(microbiomeutilities) # some utility tools 

library(broom) #tidy anova tables 
library(hrbrthemes)
library(gcookbook)
library(tidyverse)
library(dplyr)
```

#Create a vector of colors for plotting
```{r}
nice_colors = c("#999999", "#E69F00", "#56B4E9","#e98756","firebrick3","#5800e6", "#CDDC49", "#C475D3", 
                "#E94B30", "#233F57", "#FEE659", "#A1CFDD", "#F4755E", "#D6F6F7","blueviolet", "#6898BF","cornsilk",  "tan3", "violetred4", "seashell3", "forestgreen", "grey60", "hotpink", "darkgreen", "dodgerblue3", "gold", "springgreen", "slateblue2", "darkolivegreen", "deeppink4", "gray16", "white", "yellow", "red", "tan", "slategray3", "wheat", "thistle4", "peachpuff", "plum", "orangered", "mediumseagreen", "hotpink4", "goldenrod3", "ivory", "darkturquoise", "lightgoldenrod", "violet", "firebrick", "mistyrose2", "brown", "rosybrown", "royalblue", "orange", "blue", "hotpink2", "tan4", "plum2", "grey40", "firebrick4", "violet", "brown4", "ivory2", "seashell4")
                
#ASsign colors to factors
color_factor<- c("Summer" = "firebrick",
"Fall" =          "gold"  ,
"Winter" =            "dodgerblue" ,  
"Spring" =          "plum" ,
 "Raw" =           "coral"  ,     
 "Filling" =           "darkseagreen",
 "Post-Pasteurization" = "darkorchid",
"Post-Pasteurization " = "darkorchid",
"Warehouse"= "burlywood",
"Wall/Ceiling" = "mediumaquamarine",
"Door" = "grey",
"Fan/Vent" = "lightgoldenrod",
"Floor" = "hotpink",
"Floor Object" = "hotpink")
 
```


#Load Data and Convert to Phyloseq Object
```{r}
#Convert qiime artifacts directly to phyloseq
#phyloseq provides a set of classes and tools to facilitate the import, storage, analysis, and graphical display of microbiome census data.

#List files
list.files()

#Importing OTU/ASV table
ASVS <- read_qza("q20/otu-table-dn-99-SFWS-q20.qza")   #Qiime output ASV file after filtering
dim(ASVS$data)

# Importing phylogenic tree
tree <- read_qza("q20/rooted_tree.qza")  #Qiime output file
tree$data

# Importing taxonomy assignment 
taxonomy <- read_qza("q20/unite-dyn-11-22-ITS-taxonomy-otu.qza")  #Qiime output file

#Process taxonomy file more 
#get taxonomy names
tax_strings<- strsplit(as.character(taxonomy$data$Taxon), ";")
head(tax_strings)

#some strings are missing levels (blanks) and the dataframe will not bind
x = 1
y = 1

blanks <- c("k__","p__", "c__", "o__","f__", "g__", "s__")

num_taxa = length(tax_strings)  #number of taxa
num_taxa  #shoudl match dim OTUs$data

for (item in tax_strings){
  
  t = length(tax_strings[[x]]) #get list item length
  
if (t < 7)  {  #if less than 7 elements in each list item
  #print(t)
  y=1
  #print(taxa)
  while (y <= 7){  #loop through each item in item 
    if (is.na(tax_strings[[x]][y])){
      print('needs')
      tax_strings[[x]][y] = blanks[y]  #replace blank space with corr level character
      y=y+1
      }
      else(y=y+1)
    }
x = x+1
}
  else (x = x+ 1)
}

#convert to dataframe
tax_table<-as.data.frame(tax_strings)
dim(tax_table)
#Trasnpose table
tax_table<-t(tax_table)
head(tax_table)

#Add level labels
#Add oto/asv info to tax table
rownames(tax_table) <- taxonomy$data$Feature.ID
colnames(tax_table) <- c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
head(tax_table)
#Inspect table 
write.csv(tax_table, "merged_tax_table.csv")

#################Creating phyloseq object#######################################

#Note, phyloseq uses the term "OTU" loosely and since we fed it an ASV table, it is an ASV table
OTU = otu_table(as.matrix(ASVS$data), taxa_are_rows = T)
#Transpose OtU table
otu_table(OTU)<-t(otu_table(OTU))  #may need to do this again

TAX = tax_table(as.matrix(tax_table))
dim(TAX)
dim(taxa_names(OTU))
head(TAX)

# Importing metadata
metadata <- read.table("q20/Metadata_ITS2.txt", sep='\t', header=T, row.names=1, comment="")
head(metadata)
dim(metadata)
length(sample_names(OTU))
SAMPLE = sample_data(metadata)

#phylogentic tree
TREE = tree$data

#Inspect files for troubleshooting 
write.csv(SAMPLE, "phyloseq SAMPLE input.csv")
write.csv(TAX, "phyloseq TAX input.csv")
write.csv(OTU, "phyloseq OTU input.csv")

# merge the data
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)
#contains OTU table, taxonomy table, sample data, and phy tree
#check that dimensions are consistent (same number taxa setwd("..")
#getwd()between OTU, tax_table)
ps
#######################Summary stats from phyloseq object########################
#Summarizing the phyloseq object to check for feature of data
# check for features of data  
summarize_phyloseq(ps)
print_ps(ps)
head(sample_data(ps))


unique((sample_data(ps))$Freq_cat)

sample_data(ps)$Frequency<-replace(sample_data(ps)$Freq_cat, sample_data(ps)$Freq_cat=="None", "None/as needed")

sample_data(ps)$Frequency<-replace(sample_data(ps)$Freq_cat, sample_data(ps)$Freq_cat=="As needed", "None/as needed")

unique((sample_data(ps))$Freq_cat)


```

#Filter Phyloseq Object for missing or Low Abundance Genera
https://github.com/joey711/phyloseq/issues/1483 
```{r}
#####################################################
# Creating a table for number of features for each phyla
table(tax_table(ps)[, "Genus"], exclude = NULL)

# Removing the NAs and uncharacterized Phylums
ps0 <- subset_taxa(ps, !is.na(Genus) & !Genus %in% c("", "g__"))#Remove unidentified genera

# Summary Before and After
ps
ps0

# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(ps0),
               MARGIN = ifelse(taxa_are_rows(ps0), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps0),
                    tax_table(ps0))
# Viewing
head(prevdf)
# Computing average and total prevalence
temp_df <- plyr::ddply(prevdf, "Genus", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

newdata<- temp_df
colnames(newdata) <- c("Genus", "mean","sum_total")
head(newdata)

newdata<- newdata%>% arrange(desc(sum_total))
head(newdata)

write.csv(newdata, "Prevlance_Genus.csv")

# Vector to be removed
filterGenus <- temp_df[temp_df$`2` < 10,]$Genus  #Can change this number should be at least greater than 2

# Filter entries with unidentified Phylum.
ps1 = subset_taxa(ps0, !Genus %in% filterGenus)

# Summary
ps
ps0
ps1

```

#Filter samples By categories 
```{r}
#GEt samples to keep for Built Environment (n=119)
keep<- read.csv("keep.csv")
ps_be<- subset_samples(ps1, samples %in% keep$Sample.ID)
ps_be
capture.output((ps_be), file = "Phyloseq_Object_BE.txt")


#Filter by Facility 
ps_be_NL<- subset_samples(ps_be, Facility=="Facility3")
ps_be_NL
ps_be_B<- subset_samples(ps_be, Facility=="Facility2")
ps_be_B
ps_be_WS<- subset_samples(ps_be, Facility=="Facility1")
ps_be_WS

#Room
ps_PP<- subset_samples(ps_be, RoomType %in% c("Post-Pasteurization "))
ps_PP
ps_Ware<- subset_samples(ps_be, RoomType %in% c("Warehouse"))
ps_Ware
ps_Raw<- subset_samples(ps_be, RoomType %in% c("Raw"))
ps_Raw
ps_Filling<- subset_samples(ps_be, RoomType %in% c("Filling"))
ps_Filling


#Season
ps_Fall<- subset_samples(ps_be, Season=="Fall")
ps_Fall
ps_Spring<- subset_samples(ps_be, Season=="Spring")
ps_Spring
ps_Winter<- subset_samples(ps_be, Season=="Winter")
ps_Winter
ps_Summer<- subset_samples(ps_be, Season=="Summer")
ps_Summer

#oBJECT
ps_be_fan<- subset_samples(ps_be, Object =="Fan/Vent")
ps_be_fan
ps_be_wall<- subset_samples(ps_be, Object =="Wall/Ceiling")
ps_be_wall
ps_be_floor<- subset_samples(ps_be, Object =="Floor Object")
ps_be_floor
ps_be_door<- subset_samples(ps_be, Object =="Door")
ps_be_door

```


#Create Heatmap Function 
```{r}
setwd("./Heat_map")

######################################Aggretate genus
#Create Function for Heatmap plot
#df_y = aggregated dataframe
#name = Metadata category
#per = percent relative abunance to cut off 


heat_plots<- function (df_y,name,per) {

df<- as.data.frame(df_y)

#ORder Columns
df2<- df[ , order(names(df))]
head(df2)

#Remove g__
row.names(df2) <- gsub("g__", "",rownames(df2))  # no libraries needed
head(df2)
df2 <- df2[order(row.names(df2)), , drop = FALSE]
head(df2)

#Pivot data
df4 <- tibble::rownames_to_column(df2, "Genus")
head(df4)

#Aggregate Colums
colnames(df4) <- gsub("\\..*","",colnames(df4))  # no libraries needed
head(df4)

#Aggregate columnes same naem df
library(dplyr)
col <- unique(colnames(df4))[-1] 
col
df5<- cbind(df4[1], sapply(col, function(x) rowMeans(df4[,grepl(x, colnames(df4))])))

head(df5)
tail(df5)
dim(df5)

df6<- df5

#Calculate relative abundance of datafame
i = 1 #roew
j= 2 #column
while(i<= dim(df5)[1]){
    j=2
  while(j<= dim(df5)[2]){
    element = df5[i,j]
    #print(element)
    sum_col = sum(df5[,j])
    #print(sum_col)
    df6[i,j] <- (element/sum_col ) *100
    print(df6[i,j])
    
    j= j+1
  }
  
  i=i+1
}

head(df6)
tail(df6)
write.csv(df6, paste(name, per,"RA_Production.csv"))

#Then check each value< remove rows with no values > per RA
list.cols<- list()
keeps<- list()

i = 1
j = 2
while (i<= dim(df6)[1]){  #For each row
j = 2
list.cols<- list()
  while(j<= dim(df6)[2]){ #For each column
  list.cols[[j]] = df6[i,j] >= per #### change PERCENT .0.01###############
  j = j+1
  }
  
  #As long as one true keep
  vec<- unlist(list.cols)
  print(vec)
  print('new')
  if ('TRUE' %in% vec){keeps[[i]]= i}
    i = i+1
  }

keeps<- unlist(keeps)
keeps  #Rows to remove

#Fitler to only have keeps
df7<- df6 %>% slice(keeps)
dim(df7)
dim(df6)

df7

#https://jcoliver.github.io/learn-r/006-heatmaps.html

mine_long <- pivot_longer(data = (df7), 
                          cols = -c("Genus"),
                          names_to = "Site", 
                          values_to = "Abundance")
head(mine_long)
tail(mine_long)

mine_long$Site_Room<- mine_long$Site
head(mine_long)


head(mine_long)

#Fiter out "Fungi_gen_Incertae_sedis"

mine_long <- filter(mine_long, !(Genus %in% c("Fungi_gen_Incertae_sedis", "Rozellomycota_gen_Incertae_sedis")))

#PLot
mine_heatmap <- ggplot(data = mine_long, mapping = aes(x = Site,
                                                       y = Genus,
                                                       fill = Abundance)) +
  geom_tile(color = "gray") +
  #xlab(label = "Depth (m)") +
  #facet_grid(~ Site_Room, switch = "x", scales = "free_x", space = "free_x") +
    scale_fill_gradient(name = "Relative Abundance",
                      low = "grey95",  ##FFFFFF
                      high = "darkblue",
                      limits=c(0,35)) +  ##012345
  theme(strip.placement = "outside",
        plot.title = element_text(hjust = 0.5)) +
  #ggtitle(label = "Microbe Class Abundance")+
    theme(axis.text.y = element_text(color="black", size=12), axis.text.x = element_text(color="Black", size=10))+
  theme(strip.text.x = element_text(size = 10, colour = "black", angle=45))+
    scale_y_discrete(expand = c(0, 0))+ #Force axis to start at 0 
      scale_x_discrete(expand = c(0, 0)) #Force axis to start at 0 


mine_heatmap

tiff(paste(name, per," per Heatmap Genus.tif"), res=300, compression = "lzw", height=8, width=8, units="in")
print(mine_heatmap)
dev.off()

pdf(paste(name, per, " per Heatmap Genus.pdf"), height=8, width=8)
print(mine_heatmap)
dev.off()

write.csv(df7, paste(name, per, "list Production.csv"))
}
```

#Create Heatmap for Object Type
```{r}
###Create Aggredated Dataframe for Object

physeq_genus <- aggregate_taxa(ps_be, level = "Genus")

physeq_genus
otu_table(physeq_genus)
#Calculate Relative Abundance
physeq.fam.rel <- microbiome::transform(physeq_genus, "compositional")

#convert to matrix 
y<- data.matrix(otu_table(physeq.fam.rel), rownames.force = NA)
head(y)
#Match with metadata
meta<- sample_data(physeq.fam.rel)
meta$sample= rownames(meta)
head(meta)

vector<- meta$Object[match(colnames(y), meta$sample)]
vector
colnames(y)<- vector
head(y)

#ABC order
class(y)
df_o<- y
###############RUN Function ########
heat_plots(df_o, "Object", 5)  #5% cutoff
heat_plots(df_o, "Object", 1)  #1% cutoff

```

#Create heatmap for Season 
```{r}
###Season Aggreate dataframe for season 
physeq_genus
otu_table(physeq_genus)
#Calculate Relative Abundance
physeq.fam.rel <- microbiome::transform(physeq_genus, "compositional")

#convert to matrix 
y<- data.matrix(otu_table(physeq.fam.rel), rownames.force = NA)
head(y)
#Match with metadata
meta<- sample_data(physeq.fam.rel)
meta$sample= rownames(meta)
head(meta)

vector<- meta$Season[match(colnames(y), meta$sample)]
vector
colnames(y)<- vector
head(y)

#ABC order
class(y)
df_o<- y
###############
heat_plots(df_o, "Season", 5)  #%% cutoff
heat_plots(df_o, "Season", 1)  #1% cutoff
```

#Create heatmap for Room type 
```{r}
###Area##########################################################
physeq_genus
otu_table(physeq_genus)
#Calculate Relative Abundance
physeq.fam.rel <- microbiome::transform(physeq_genus, "compositional")

#convert to matrix 
y<- data.matrix(otu_table(physeq.fam.rel), rownames.force = NA)
head(y)
#Match with metadata
meta<- sample_data(physeq.fam.rel)
meta$sample= rownames(meta)
head(meta)

vector<- meta$RoomType[match(colnames(y), meta$sample)]
vector
colnames(y)<- vector
head(y)

#ABC order
class(y)
df_o<- y
###############
heat_plots(df_o, "RoomType", 5) 
  heat_plots(df_o, "RoomType", 1)

#Reset directory 
setwd("..")
getwd()

```

#Beta Diversity PCOA Plots
```{r}
####################################################################################
#FUNCTION 
#calculate Beta diversity wtih phyloseq object
#name  is a "string" to identify titles 
setwd("./Beta_Div_area")

beta_diversity_phylogenic <- function(ps_object, meth, color_var){

set.seed(101)

physeq.ord.bray <- ordinate(ps_object, "PCoA", "bray", weighted=F)
b.div.bray <- plot_ordination(ps_object, physeq.ord.bray , type= "samples", color= "RoomType") + geom_point(size=18)
#+scale_color_manual(values=mycolors)

b.div.bray <- b.div.bray  + ggtitle("Bray")  + theme_classic() + theme(axis.text.x = element_text(color= "black", size = 25, angle=45),axis.text.y = element_text(color= "black", size = 25), text = element_text(size=25,colour="black",face="bold"),strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))+ stat_ellipse(geom = "polygon", alpha = 0.1, aes(color= RoomType, fill = RoomType))+
scale_color_manual(values=color_factor)+scale_fill_manual(values=color_factor)
#


print(b.div.bray)

#Export plot to .tif file 
tiff(paste(meth, color_var,"Bray Beta Diversity.tif", sep=" "), res=300, compression = "lzw", height=9, width=12, units="in")
print(b.div.bray)
dev.off()

pdf(paste(meth, color_var,"Bray Beta Diversity.pdf", sep=" "), height=9, width=12)
print(b.div.bray)
dev.off()

}
####################################################################################
#Call Function 
#Filter ps_be samples to get rid of zero rows

#Get ride of Zero Rows or Beta Diversity will not work.  
ps_be_f <- subset_samples(ps_be, !samples %in% c("sample-11", "sample-110", "sample-133", "sample-135", "sample-137", "sample-138", "sample-346" ,"sample-347", "sample-353", "sample-247", "sample-331","sample-332"))

ps_be_fan2<- subset_samples(ps_be_f, Object =="Fan/Vent")
ps_be_wall2<- subset_samples(ps_be_f, Object =="Wall/Ceiling")
ps_be_floor2<- subset_samples(ps_be_f, Object =="Floor Object")
ps_be_door2<- subset_samples(ps_be_f, Object =="Door")

###########################################################
beta_diversity_phylogenic(ps_be_fan2, "Fan", "RoomType_Color")
beta_diversity_phylogenic(ps_be_floor2, "Floor", "RoomType_Color")
beta_diversity_phylogenic(ps_be_door2, "Door", "RoomType_Color")
beta_diversity_phylogenic(ps_be_wall2, "Wall", "RoomType_color")

#beta_diversity_phylogenic(ps_be_f, "All", "Season")


setwd("..")
getwd()

```



#Beta Diversity PCOA Plots-ALL
```{r}
####################################################################################
#FUNCTION 
#calculate Beta diversity wtih phyloseq object
#name  is a "string" to identify titles 
setwd("./Beta_All_pcoa")

beta_diversity_phylogenic <- function(ps_object, meth, color_var){

set.seed(101)

#mycolors = c("Filling" = "red", "Warehouse"="blue", "Post-Pasteurization"="green", "Raw" = "violet")
#Order rooms
#sample_data(ps_object)$RoomType <- ordered(sample_data(ps_object)$RoomType, levels =c("Raw", "Filling", "Post-Pasteurization", "Warehouse"))

  
physeq.ord.bray <- ordinate(ps_object, "PCoA", "bray", weighted=F)
b.div.bray <- plot_ordination(ps_object, physeq.ord.bray , type= "samples", color= color_var) + geom_point(size=18)
#+scale_color_manual(values=mycolors)

b.div.bray <- b.div.bray  + ggtitle("Bray")  + theme_classic() + theme(axis.text.x = element_text(color= "black", size = 25),axis.text.y = element_text(color= "black", size = 25), text = element_text(size=25,colour="black",face="bold"),strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))+ stat_ellipse(geom = "polygon", alpha = 0.1, aes(color= Facility, fill = Facility))
#+scale_color_manual(values=mycolors)
#


print(b.div.bray)

#Export plot to .tif file 
tiff(paste(meth, color_var,"Bray Beta Diversity.tif", sep=" "), res=300, compression = "lzw", height=9, width=15, units="in")
print(b.div.bray)
dev.off()

pdf(paste(meth, color_var,"Bray Beta Diversity.pdf", sep=" "), height=9, width=15)
print(b.div.bray)
dev.off()

}
####################################################################################
#Call Function 
#Filter ps_be samples to get rid of zero rows
#keep_samples <- apply(otu_table(ps_be), 1, function(x) paste(names(otu_table(ps_be))[x == 0], collapse = "-"))
#d<- as.data.frame(keep_samples)
#head(d)

#Get ride of Zero Rows or Beta Diversity will not work.  
ps_be_f <- subset_samples(ps_be, !samples %in% c("sample-11", "sample-110", "sample-133", "sample-135", "sample-137", "sample-138", "sample-346" ,"sample-347", "sample-353", "sample-247", "sample-331","sample-332"))


###########################################################

beta_diversity_phylogenic(ps_be_f, "All", "Facility")


setwd("..")
getwd()

```


#Permanova-Beta diversity 
```{r}
setwd("./Beta_permanova_object")

#install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)

#Process otu files and metadata
otu <- abundances(ps_be) 
meta <- meta(ps_be)

#Clean data
r = 1 #rows
c= 1 #colcums
#Add pseudocounts
#print(dim(otu))
row_n = dim(otu)[1]
col_n = dim(otu)[2]
while (r <= row_n){
  c=1
  while (c<=col_n){
  if(otu[r,c]==0){
    print('yes')
    otu[r,c]=1}
    c=c+1
  }
  r=r+1
}
#Calculate bray curtis distance
dist <- vegdist(t(otu), "bray")
print(dist)

#Remove "inf from distnce
len = length(dist)
#Add pseudocounts
#print(len)
j = 1
while (j <= len){
    if(dist[j]==Inf || dist[j] ==-Inf){
    print('yes')
    dist[j]=0}
    j=j+1
}


################################################################################
beta_disp <- function(name){
set.seed(99)
#Create betadispirs plot
bd = betadisper(dist, group = meta[,c(name)])

df <- data.frame(Distance_to_centroid=bd$distances,Group=bd$group)
groups <- bd$group

mod <- betadisper(dist, groups)

#Anova
aov<- anova(mod)
write.csv(aov, paste(name,"anova_mod_BRAY.csv"))

## Tukey's Honest Significant Differences
(mod.HSD <- TukeyHSD(mod))
plot(mod.HSD)
print(mod.HSD)
write.csv(mod.HSD$group, paste(name, "Tukey BRAY.csv"))


print(mod)

#https://github.com/vegandevs/vegan/issues/166
#The intention of betadisper is indeed to compare only sizes of ellipses. These sizes (covariances) will be calculated with respect to the centres used, and these centres default to spatial medians. These are somewhat wider than covariances calculated with respect to sample means. Therefore the centres are also important. We are maintaining the options to plot SD ellipses, but you must set type="centroid" to get the ellipses of correct sizes for the plot. Jury is still out for the modification of SD's that you seem to ask for.

library(vegan)
iris.bd <- wcmdscale(dist, eig = TRUE)  # PCoA

#pairwise comparisions 
post_hoc_permanova <- pairwise.adonis(t(otu), meta[,c(name)], sim.function = "vegdist", 
sim.method = "bray", p.adjust.m = "fdr", reduce = NULL, perm = 999) 
post_hoc_permanova 
write.csv(post_hoc_permanova, paste(name,"posthoc_BRAY.csv"))


}

############################################################################
beta_disp("Season")
beta_disp("Object")
beta_disp("Facility")
beta_disp("RoomType")



#PERMANOVA-Adonis
set.seed(92)
#https://uw.pressbooks.pub/appliedmultivariatestatistics/chapter/permanova/ 

#Y ~ a + b + a:b means that the response is modeled by a linear combination of a, b, and the interaction between a and b
ad<- adonis2(dist ~meta$Facility+meta$RoomType+meta$Object+meta$Season+meta$Facility:meta$RoomType:meta$Object:meta$Season, method = "bray")
ad
write.csv(ad, "All_factors_nested.csv")

ad<- adonis2(dist ~meta$Facility+meta$RoomType+meta$Object+meta$Season+meta$Facility:meta$RoomType+ +meta$Season+meta$Facility:meta$RoomType:meta$Object+meta$Season+meta$Facility:meta$RoomType:meta$Object:meta$Season, method = "bray")
ad
write.csv(ad, "All_factors_nested_all_combins.csv")

#*
ad<- adonis2(dist ~meta$Facility*meta$RoomType*meta$Object*meta$Season, method = "bray")
ad
write.csv(ad, "All_factors_nested_all_combins_sar.csv")



#Remove Object-use when subetting by object 
ad<- adonis2(dist ~meta$Facility+meta$RoomType+meta$Season+meta$RoomType:meta$Facility:meta$Season, method = "bray")
ad
write.csv(ad, "All_factors_nested2.csv")

#OVerall permanova
Facility<- adonis2(dist ~meta$Facility, method = "bray")
write.csv(Facility, "fACILITY PERMANOVA.CSV")

AR<- adonis2(dist ~meta$RoomType, method = "bray")
AR
write.csv(AR, "AREA PERMANOVA.CSV")
Season<- adonis2(dist ~meta$Season, method = "bray")
Season

ob<- adonis2(dist ~meta$Object, method = "bray")
ob
write.csv(ob, "Object PERMANOVA.CSV")


#################################
#Individual plots--SEASON

set.seed(99)
#Create betadispirs plot
bd = betadisper(dist, group = meta$Season)

df <- data.frame(Distance_to_centroid=bd$distances,Group=bd$group)
groups <- bd$group
df
mod <- betadisper(dist, groups)
mod
#https://github.com/vegandevs/vegan/issues/166
#The intention of betadisper is indeed to compare only sizes of ellipses. These sizes (covariances) will be calculated with respect to the centres used, and these centres default to spatial medians. These are somewhat wider than covariances calculated with respect to sample means. Therefore the centres are also important. We are maintaining the options to plot SD ellipses, but you must set type="centroid" to get the ellipses of correct sizes for the plot. Jury is still out for the modification of SD's that you seem to ask for.

library(vegan)
iris.bd <- wcmdscale(dist, eig = TRUE)  # PCoA
x_axis = paste("PCOA1 (", round(iris.bd$eig[[1]],2), "%)")
y_axis = paste("PCOA1 (", round(iris.bd$eig[[2]],2), "%)")

#Save as .tiff
library(heplots)

labs <- paste0("Dimension ", 1:4, " (", 
              round(100*iris.bd$eig / sum(iris.bd$eig), 2), "%)")

#Get point values
xx<- as.data.frame(iris.bd$points[,1:2])
head(xx)
xx$sampleid<- rownames(xx)
head(xx)
xx$Season<- metadata$Season[match(xx$sampleid, metadata$samples)]
xx$Prod<- metadata$RoomType[match(xx$sampleid, metadata$samples)]
head(xx)




###############################SEASON################
plot <- ggplot(xx, aes(x=Dim1,y=Dim2)) +
      geom_point(aes(fill=Season,stroke=NA),pch=21,size=3) +
      stat_ellipse(geom = "polygon",aes(group=Season,color=Season,fill=Season),alpha = 0.25, color=NA) +
     scale_fill_manual(values=color_factor)+
theme(text = element_text(size=20,colour="black",face="bold"), 
       axis.text.x = element_text(angle=45,hjust = 1, size=25, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.text.y = element_text(angle = 0, hjust = 1, size=25, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.ticks.length = unit(-0.25, "cm"), #-0.25 (inside)
     plot.title= element_text(hjust = 0.9))+ theme(panel.background = element_blank(),axis.line = element_line( colour = "black"))+
  labs(x = x_axis, y = y_axis) +
    theme(axis.title=element_text(size=25, face="bold"))


tiff(filename=paste("SEASON betadispr_color.tiff"), width=12, height=9, bg="white", res=300, units = "in")
plot(plot)
dev.off()

pdf(paste("SEASON betadispr_color.pdf", sep=" "), width=12, height=9)
plot(plot)
dev.off()

###############################AREA###############
plot <- ggplot(xx, aes(x=Dim1,y=Dim2)) +
      geom_point(aes(fill=Prod,stroke=NA),pch=21,size=3) +
      stat_ellipse(geom = "polygon",aes(group=Prod,color=Season,fill=Prod),alpha = 0.25, color=NA) +
     scale_fill_manual(values=color_factor)+
theme(text = element_text(size=20,colour="black",face="bold"), 
       axis.text.x = element_text(angle=45,hjust = 1, size=25, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.text.y = element_text(angle = 0, hjust = 1, size=25,colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.ticks.length = unit(-0.25, "cm"), #-0.25 (inside)
     plot.title= element_text(hjust = 0.9))+ theme(panel.background = element_blank(),axis.line = element_line( colour = "black"))+
  labs(x = x_axis, y = y_axis) +
  theme(axis.title=element_text(size=25, face="bold"))


tiff(filename=paste("PRODUCTION AREA betadispr_color.tiff"), width=12, height=9, bg="white", res=300, units = "in")
plot(plot)
dev.off()

pdf(paste("PRODUCTION AREA betadispr_color.pdf", sep=" "), width=12, height=9)
plot(plot)
dev.off()

###############################OBJECT###############

set.seed(99)
#Create betadispirs plot
bd = betadisper(dist, group = meta$Object)

df <- data.frame(Distance_to_centroid=bd$distances,Group=bd$group)
groups <- bd$group
df
mod <- betadisper(dist, groups)
mod
#https://github.com/vegandevs/vegan/issues/166
#The intention of betadisper is indeed to compare only sizes of ellipses. These sizes (covariances) will be calculated with respect to the centres used, and these centres default to spatial medians. These are somewhat wider than covariances calculated with respect to sample means. Therefore the centres are also important. We are maintaining the options to plot SD ellipses, but you must set type="centroid" to get the ellipses of correct sizes for the plot. Jury is still out for the modification of SD's that you seem to ask for.

library(vegan)
iris.bd <- wcmdscale(dist, eig = TRUE)  # PCoA
x_axis = paste("PCOA1 (", round(iris.bd$eig[[1]],2), "%)")
y_axis = paste("PCOA1 (", round(iris.bd$eig[[2]],2), "%)")

#Save as .tiff
library(heplots)

labs <- paste0("Dimension ", 1:4, " (", 
              round(100*iris.bd$eig / sum(iris.bd$eig), 2), "%)")

#Get point values
xx<- as.data.frame(iris.bd$points[,1:2])
head(xx)
xx$sampleid<- rownames(xx)
head(xx)
xx$Object<- metadata$Object[match(xx$sampleid, metadata$samples)]
xx$Season<- metadata$Season[match(xx$sampleid, metadata$samples)]
xx$Prod<- metadata$RoomType[match(xx$sampleid, metadata$samples)]
head(xx)

plot <- ggplot(xx, aes(x=Dim1,y=Dim2)) +
      geom_point(aes(fill=Object,stroke=NA),pch=21,size=3) +
      stat_ellipse(geom = "polygon",aes(group=Object,color=Object,fill=Object),alpha = 0.25, color=NA) +
     scale_fill_manual(values=color_factor)+
theme(text = element_text(size=20,colour="black",face="bold"), 
       axis.text.x = element_text(angle=45,hjust = 1, size=25, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.text.y = element_text(angle = 0, hjust = 1, size=25,colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.ticks.length = unit(-0.25, "cm"), #-0.25 (inside)
     plot.title= element_text(hjust = 0.9))+ theme(panel.background = element_blank(),axis.line = element_line( colour = "black"))+
  labs(x = x_axis, y = y_axis) +
  theme(axis.title=element_text(size=25, face="bold"))

plot

tiff(filename=paste("OBJECT betadispr_color.tiff"), width=12, height=9, bg="white", res=300, units = "in")
plot(plot)
dev.off()

pdf(paste("OBJECT betadispr_color.pdf", sep=" "), width=12, height=9)
plot(plot)
dev.off()

setwd("..")
getwd()

```

#Alpha Diversity 
```{r}
###############################################################
#CALL Function 
#alpha_=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson")

setwd("./Alpha_div")
#Get Richness Data
alpha.div2 <- plot_richness(ps, x="RoomType", measures="Chao1") +
    geom_boxplot() +
    theme_classic() +
    theme(text = element_text(size=24,colour="black",face="bold"),strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))

alpha.div2
head(alpha.div2$data)

write.csv(alpha.div2$data, "Chao1 Diversity.csv")


#Boxplot wtih distribiton
#Wdith = probablity point will be that value --wider means greater probabliity 
alpha_div_stats <- plot_diversity_stats(ps_be, group = "Season", 
                     index = "diversity_shannon",
                     group.order = c("Summer","Fall","Winter", "Spring"),                      
                     group.colors = c("firebrick", "gold","dodgerblue", "plum"))+
                     #label.format="p.format",stats = TRUE))+
                    ylab("Shannon Diversity") + xlab("Season")+
    scale_y_continuous(expand = c(0, 0), limits=c(0,5), breaks=seq(0,7, by=1))+ #Force axis to start at 0 
  theme(text = element_text(size=20,colour="black",face="bold"), 
       axis.text.x = element_text(angle=45,hjust = 0.5, size=20, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.text.y = element_text(angle = 0, hjust = 1, size=25, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.ticks.length = unit(-0.25, "cm"))+ #-0.25 (inside)
    scale_x_discrete(expand = c(0, 0))
    

print(alpha_div_stats)

tiff(paste("SEASON Shannon Boxplot.tif", sep=" "), res=300, compression = "lzw", height=9, width=9, units="in")
print(alpha_div_stats)
dev.off()

pdf(paste("SEASON Shannon Boxplot.pdf", sep=" "), height=9, width=9)
print(alpha_div_stats)
dev.off()


#####################################
alpha_div_statsR <- plot_diversity_stats(ps_be, group = "RoomType", 
                     index = "diversity_shannon",
                     group.order = c("Raw", "Filling","Post-Pasteurization ", "Warehouse"),                      
                     group.colors = c( "coral", "darkseagreen", "darkorchid", "burlywood"),
                     label.format=NULL)+
                    ylab("Shannon Diversity") + xlab("Season")+
  scale_y_continuous(expand = c(0, 0), limits=c(0,5) breaks=seq(0,7, by=1))+ #Force axis to start at 0 
  theme(text = element_text(size=25,colour="black",face="bold"),
       axis.text.x = element_text(angle=45,hjust = 0.5, size=18, colour = 'black', "cm"),
     axis.text.y = element_text(angle = 0, hjust = 1, colour = 'black', size=25, margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.ticks.length = unit(-0.25, "cm"))+ #-0.25 (inside)
     scale_x_discrete(expand = c(0, 0))

print(alpha_div_statsR)

tiff(paste("Shannon Boxplot Room.tif", sep=" "), res=300, compression = "lzw", height=9, width=9, units="in")
print(alpha_div_statsR)
dev.off()

pdf(paste("Shannon Boxplot Room.pdf"), height=9, width=9)
print(alpha_div_statsR)
dev.off()

#######################OBJECT
#####################################

alpha_div_statsO <- plot_diversity_stats(ps_be, group = "Object", 
                     index = "diversity_shannon",
                     group.order = c("Door", "Fan/Vent", "Floor Object", "Wall/Ceiling"),                      
                     group.colors = c( "grey","lightgoldenrod", "hotpink", "mediumaquamarine"),
                     label.format=NULL)+
                    ylab("Shannon Diversity") + xlab("Object")+
  scale_y_continuous(expand = c(0, 0), limits=c(0,5), breaks=seq(0,7, by=1))+ #Force axis to start at 0 
  theme(text = element_text(size=25,colour="black",face="bold"),
       axis.text.x = element_text(angle=45,hjust = 0.5, size=18, colour = 'black', "cm"),
     axis.text.y = element_text(angle = 0, hjust = 1, colour = 'black', size=25, margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.ticks.length = unit(-0.25, "cm"))+ #-0.25 (inside)
     scale_x_discrete(expand = c(0, 0))

print(alpha_div_statsO)

tiff(paste("Shannon Boxplot Object.tif", sep=" "), res=300, compression = "lzw", height=9, width=9, units="in")
print(alpha_div_statsO)
dev.off()

pdf(paste("Shannon Boxplot Object.pdf"), height=9, width=9)
print(alpha_div_statsO)
dev.off()


setwd("..")
getwd()

```

#Tornado PLot-By Facility 
```{r}
#FUNCTION###########################################################################
#PS_object = Phyloseq object
#name = File name -does not affect analysis - (string)

List_taxa <- function(ps_object, name){
setwd(("./tornado_plot_Facility"))
grp_abund <- get_group_abundances(ps_object, 
                                  level = "Genus", 
                                  group=NULL,
                                  transform = "compositional")

grp_abund$OTUID <- gsub("p__", "",grp_abund$OTUID)
grp_abund$OTUID <- ifelse(grp_abund$OTUID == "", 
                          "Unclassified", grp_abund$OTUID)


write.csv(grp_abund, paste(name, "Mean Abundance Genus.csv", sep=" "))
  
#filter out unidentified genus
abund_df = grp_abund

#sort abund_df by mean
abund_df = as.data.frame(abund_df)
newdata <- abund_df[order(abund_df$mean_abundance, decreasing = TRUE),]
newdata[1:10,]

write.csv (newdata[1:10,],paste(name,"Top Ten Genus .csv", sep = " "))

ten_df <- as.data.frame(newdata[1:10,])  # input data

#clean strings
i=1
while (i <= dim(ten_df)[1]){
#If string starts with "k"
if (substr(ten_df[i,1],1,1)=="k") {
  ten_df[i,1] =sapply(str_split(ten_df[i,1],"f__",),'[',2)  #split at order 
}
else  
ten_df[i,1]=sapply(str_split(ten_df[i,1],"g__",),'[',2)  #Remove g from genus 
i=i+1
}

ten_df$mean_abundance = 100 * ten_df$mean_abundance
print(ten_df)

bar <-ggplot(ten_df, aes(x=reorder(OTUID,-mean_abundance), y=mean_abundance)) + 
geom_bar(stat = "identity", color= "deepskyblue2", fill = "deepskyblue2")+
theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), #Remove background
panel.background = element_blank(), axis.line = element_line(colour = "black"))+
theme(text = element_text(size=40,colour="black",face="bold"), 
       axis.text.x = element_text(angle=45,hjust = 1, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.text.y = element_text(angle = 0, hjust = 1, colour = 'black',margin=unit(c(0.5,0.5,0.5,0.5), "cm")),
     axis.ticks.length = unit(-0.25, "cm"), #-0.25 (inside)
     plot.title= element_text(hjust = 0.9))+
  scale_x_discrete(expand = c(0, 0))+
  scale_y_continuous(expand = c(0, 0))+ #Force axis to start at 0 
  #ggtitle((name))+
   ylab(expression(bold("Relative Abundance (%)")))+ #change y label
  xlab(expression(bold("Genera")))+
  theme(axis.ticks.x = element_blank())+ # Remove x axis tick marks
  coord_flip() # rotate plot 

print(bar)

#Export plot to .tif file 
tiff(paste(name, "Mean Compositional Abundance Genus.tif", sep=" "), res=300, compression = "lzw", height=9, width=15, units="in")
print(bar)
dev.off()

setwd("..")
getwd()
}
####################################################################################
#CALL Function 
List_taxa(ps_be_B, "Built Envrionment_Facility2")
List_taxa(ps_be_NL, "Built Envrionment_Facility3")
List_taxa(ps_be_WS, "Built Envrionment_Facility1")

List_taxa(ps_PP, "Built Envrionment_Post_past")
List_taxa(ps_Ware, "Built Envrionment_Ware")
List_taxa(ps_Raw, "Built Envrionment_Raw")
List_taxa(ps_Filling, "Built Envrionment_Filling")


List_taxa(ps_Summer, "Built Envrionment_SUMMER")
List_taxa(ps_Fall, "Built Envrionment_FALL")
List_taxa(ps_Winter, "Built Envrionment_WINTER")
List_taxa(ps_Spring, "Built Envrionment_SPRING")


List_taxa(ps_be_door, "Built Envrionment_DOOR")
List_taxa(ps_be_floor, "Built Envrionment_FLOOR")
List_taxa(ps_be_wall, "Built Envrionment_WALL")
List_taxa(ps_be_fan, "Built Envrionment_FAN")
```



#Differential abundance - Color by Genera 
#DeSeq2 to test for differential abundance between categories
https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/DESeq
```{r}
#Convert phyloseq object ot DeSeq

setwd(("./Diff_abundance")) 


#https://rdrr.io/github/umerijaz/microbiomeSeq/man/differential_abundance.html
physeq <- taxa_level(ps_be,"Genus")
physeq
sample_data(physeq)$Facility<- as.factor(sample_data(physeq)$Facility) #must convert to factor
deseq_sig  <- differential_abundance(physeq, grouping_column = "Facility", pvalue.threshold = 0.05)
#plot the significant features
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}

tiff("Genera_DesQ_barchart_Factor_Built Environment p0.01.tiff", res=300, compression = "lzw", height=9, width=15, units="in")
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}
dev.off()


write.csv(deseq_sig$plotdata, "Facility_deq.csv")




###############Season 
#https://rdrr.io/github/umerijaz/microbiomeSeq/man/differential_abundance.html
physeq <- taxa_level(ps_be,"Genus")
physeq
sample_data(physeq)$Season<- as.factor(sample_data(physeq)$Season) #must convert to factor
deseq_sig  <- differential_abundance(physeq, grouping_column = "Season", pvalue.threshold = 0.05)
#plot the significant features
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}

tiff("SEASON_Genera_DesQ_barchart_Built Environment p0.01.tiff", res=300, compression = "lzw", height=9, width=15, units="in")
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}
dev.off()

write.csv(deseq_sig$plotdata, "Season_deq.csv")

########################Production area
###############Season 
#https://rdrr.io/github/umerijaz/microbiomeSeq/man/differential_abundance.html
physeq <- taxa_level(ps_be,"Genus")
physeq
sample_data(physeq)$RoomType<- as.factor(sample_data(physeq)$RoomType) #must convert to factor

deseq_sig  <- differential_abundance(physeq, grouping_column = "RoomType", pvalue.threshold = 0.05)
#plot the significant features
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}

tiff("ROOM_Genera_DesQ_barchart_Built Environment p0.01.tiff", res=300, compression = "lzw", height=9, width=20, units="in")
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}
dev.off()

write.csv(deseq_sig$plotdata, "ROOM_deq.csv")


###############Object
#https://rdrr.io/github/umerijaz/microbiomeSeq/man/differential_abundance.html
physeq <- taxa_level(ps_be,"Genus")
physeq

sample_data(physeq)$Object[sample_data(physeq)$Object == "Wall/Ceiling"] <- "WallCeiling"
sample_data(physeq)$Object[sample_data(physeq)$Object == "Fan/Vent"] <- "FanVent"
sample_data(physeq)$Object[sample_data(physeq)$Object == "Floor Object"] <- "Floor"

unique(sample_data(physeq)$Object)

sample_data(physeq)$Object<- as.factor(sample_data(physeq)$Object) #must convert to factor

deseq_sig  <- differential_abundance(physeq, grouping_column = "Object", pvalue.threshold = 0.05)
#plot the significant features
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}

tiff("OBJECT_Genera_DesQ_barchart_Built Environment p0.01.tiff", res=300, compression = "lzw", height=9, width=20, units="in")
plot_signif(deseq_sig$plotdata) #see function \link[microbiomeSeq]{plot_signif}
dev.off()

write.csv(deseq_sig$plotdata, "OBJECT_deq.csv")




setwd("..")
getwd()


```

#Create Color_vector
```{r}
#I copied and pasted the results of genus_color$cat
#I cannot figure out how to concatenate the columns wtih an equal sign between them
#There are in ABC order so if there is a missing genera just add it manually to this list.
#scale_fill_manual(values=color_col) in the function below 
color_col<- c("_g__" = "slateblue2",
"Acremonium" =          "skyblue3"  ,
"Akanthomyces" =            "firebrick" ,  
"Alternaria" =          "darkturquoise" ,
 "Apiotrichum" =           "tan3"  ,     
 "Aspergillus" =           "yellow" ,    
"Aureobasidium" =             "peachpuff"  ,   
"Bisifusarium" =            "tan2" ,    
"Botryotinia" = "lemonchiffon",
"Bullera" = "aliceblue",
"Candida"  =        "red",
"Cephaloascus" =            "honeydew" ,   
"Chaetomium" =          "orange",    
"Chaetothyriales_gen_Incertae_sedis" = "tomato2",                           
"Chordomyces" =           "purple4" ,
"Cladosporium" =            "darkgreen" ,
 "Clavispora" =          "rosybrown" ,
 "Coniochaeta" =           "plum2"     , 
 "Cryptococcus" =            "yellow"  ,    
 "Curvibasidium" =             "thistle1"     ,
 "Cutaneotrichosporon" =                   "purple",          
 "Cyberlindnera" =             "mistyrose2" ,  
"Cyphellophora" =             "mintcream"   , 
"Cystobasidium" =             "white"    ,    
"Cystofilobasidium" =                 "gold2"    ,         
 "Debaryomyces"   =              "indianred4",
"Dictyosporiaceae_gen_Incertae_sedis" = "darkolivegreen1"        ,            
"Didymella" =         "skyblue"  ,
"Didymellaceae_g__" =                 "hotpink"   ,       
 "Dioszegia" =         "brown"   , 
"Diutina" =       "tan2"   ,
"Effuseotrichosporon" =                   "blue" ,              
 "Engyodontium" =            "gold3"  ,  
 "Exophiala"  = "gold4",
 "Filobasidium" =            "plum" ,       
"Fusarium" =        "wheat2",
 "Fusicolla" =         "cornsilk2",
 "Geotrichum" =          "purple4"  , 
 "Galactomyces" = "darksalmon",
 "Geotrichum/Galactomyces" = "seagreen",
 "Hanseniaspora (Kloechera)"= "violetred",
 "Humicola"   =          "indianred3",
 "Iodophanus" =          "slateblue4",
 "Kluyveromyces" =             "cadetblue" ,   
 "Knufia" =      "wheat2",
 "Kurtzmaniella" =             "olivedrab1" ,  
"Leucosporidium" =              "yellowgreen",   
 "Malassezia" =         "lightpink4",
"Meyerozyma" =         "violetred4",
 "Mrakia" =     "sienna",
 "Mucor"      =  "dodgerblue3",
"Naganishia"  =   "grey80",
"Nakazawaea" =          "peachpuff3",
"Neoascochyta" = "lightpink2",
"None" = "grey4",
"Paecilomyces" = "ivory",
 "Ophiostoma" =          "seashell4" ,
"Orbilia" =      "gold"   ,
 "Papiliotrema"   =              "mediumseagreen",
 "Penicillium " =          "springgreen" ,
 "Penicillium" =          "springgreen" ,
"Phlebiopsis"  =            "springgreen2",
 "Pichia" =      "navy",
 "Piskurozyma" =           "firebrick1" ,
 "Preussia"   =          "violetred2",
 "Proceropycnis" =             "seashell2",    
 "Rhodosporidiobolus" =                  "thistle4",
  "Rhizopus" = "darkolivegreen",
"Rhodotorula" =           "grey16",     
 "Rozellomycota_gen_Incertae_sedis" = "lemonchiffon",                    
 "Saccharomyces"  =              "paleturquoise1",
"Saccharomycetales_fam_Incertae_sedis_g__" = "grey50",  
"Sarocladium" = "slategray1",
 "Schizothecium" =             "slategray2",   
 "Scopulariopsis" =              "ivory",         
 "Stagonosporopsis" =                "hotpink3", 
 "Sporobolomyces" = "lemonchiffon2",
"Symmetrospora"  =             "lightgoldenrod",
 "Talaromyces" =           "violet",     
"Tausonia"    =           "dodgerblue1",
 "Tetracladium" =            "ivory2",   
 "Torulaspora" =           "slateblue1", 
"Trichoderma" =           "grey80",     
"Trichosporon" =           "forestgreen",
"Trametes" = "grey68",
"Uwebraunia" =          "mistyrose", 
"Valsaceae_g__" =             "thistle3",     
"Vishniacozyma" =             "deeppink4",    
 "Wallemia" =        "navy",    
"Wickerhamiella" =              "khaki",         
"Wickerhamomyces" =               "red",            
 "Yarrowia"   =          "violet")

#In your GGPLOt code use this line to get consistant colors between barcharts:  
#scale_fill_manual(values=color_col)

```

#heat_map-Production Area
```{r}
setwd("./Heat_map")

######################################Aggretate genus
heat_plots<- function (df_y,name) {
physeq_genus <- aggregate_taxa(ps_be, level = "Genus")

df<- as.data.frame(df_y)
library(dplyr)

#ORder Columns
df2<- df[ , order(names(df))]
head(df2)

#Remove g__
row.names(df2) <- gsub("g__", "",rownames(df2))  # no libraries needed
head(df2)
df2 <- df2[order(row.names(df2)), , drop = FALSE]
head(df2)


#Pivot data
df4 <- tibble::rownames_to_column(df2, "Genus")
head(df4)

#Aggregate Colums
colnames(df4) <- gsub("\\..*","",colnames(df4))  # no libraries needed
head(df4)


#Aggregate columnes same naem df
library(dplyr)
col <- unique(colnames(df4))[-1] 
col
#df5<- cbind(df4[1], sapply(col, function(x) rowSums(df4[,grepl(x, colnames(df4))])))
df5<- cbind(df4[1], sapply(col, function(x) rowMeans(df4[,grepl(x, colnames(df4))])))

head(df5)
tail(df5)
dim(df5)

df6<- df5


#Calculate relative abundance of datafame
i = 1 #roew
j= 2 #column
while(i<= dim(df5)[1]){
    j=2
  while(j<= dim(df5)[2]){
    element = df5[i,j]
    #print(element)
    sum_col = sum(df5[,j])
    #print(sum_col)
    df6[i,j] <- (element/sum_col ) *100
    print(df6[i,j])
    
    j= j+1
  }
  
  i=i+1
}

head(df6)
tail(df6)
write.csv(df6, "RA_Production.csv")

#Then check each value< remove rows with no values > 5% RA
list.cols<- list()
keeps<- list()

i = 1
j = 2
while (i<= dim(df6)[1]){  #For each row
j = 2
list.cols<- list()
  while(j<= dim(df6)[2]){ #For each column
  list.cols[[j]] = df6[i,j] >= 1  ####chANGE PERCETN HERE 
  j = j+1
  }
  
  #As long as one true keep
  vec<- unlist(list.cols)
  print(vec)
  print('new')
  if ('TRUE' %in% vec){keeps[[i]]= i}
  #else{keeps[[i]]== 'drop'}
    i = i+1
  }

keeps<- unlist(keeps)
keeps  #Rows to remove

#Fitler to only have keeps
df7<- df6 %>% slice(keeps)
dim(df7)
dim(df6)

df7

#https://jcoliver.github.io/learn-r/006-heatmaps.html

mine_long <- pivot_longer(data = (df7), 
                          cols = -c("Genus"),
                          names_to = "Site", 
                          values_to = "Abundance")
head(mine_long)
tail(mine_long)

mine_long$Site_Room<- mine_long$Site
head(mine_long)


head(mine_long)

#Fiter out "Fungi_gen_Incertae_sedis"

mine_long <- filter(mine_long, !(Genus %in% "Fungi_gen_Incertae_sedis"))

#PLot
mine_heatmap <- ggplot(data = mine_long, mapping = aes(x = Site,
                                                       y = Genus,
                                                       fill = Abundance)) +
  geom_tile(color = "gray") +
  #xlab(label = "Depth (m)") +
  #facet_grid(~ Site_Room, switch = "x", scales = "free_x", space = "free_x") +
    scale_fill_gradient(name = "Relative Abundance",
                      low = "grey95",  ##FFFFFF
                      high = "darkblue",
                      limits=c(0,35)) +  ##012345
  theme(strip.placement = "outside",
        plot.title = element_text(hjust = 0.5)) +
  #ggtitle(label = "Microbe Class Abundance")+
    theme(axis.text.y = element_text(color="black", size=12), axis.text.x = element_text(color="Black", size=12))+
  theme(strip.text.x = element_text(size = 12, colour = "black"))+
    scale_y_discrete(expand = c(0, 0))+ #Force axis to start at 0 
      scale_x_discrete(expand = c(0, 0)) #Force axis to start at 0 


mine_heatmap

print(mine_heatmap)
dev.off()

pdf(paste(name,"1 per Proudction Heatmap Genus.pdf"), height=8, width=8)
print(mine_heatmap)
dev.off()

write.csv(df7, paste(name,"Season list Production.csv"))
}

##############################OBECT#############################

###Object
physeq_genus
otu_table(physeq_genus)
#Calculate Relative Abundance
physeq.fam.rel <- microbiome::transform(physeq_genus, "compositional")

#convert to matrix 
y<- data.matrix(otu_table(physeq.fam.rel), rownames.force = NA)
head(y)
#Match with metadata
meta<- sample_data(physeq.fam.rel)
meta$sample= rownames(meta)
head(meta)

vector<- meta$Object[match(colnames(y), meta$sample)]
vector
colnames(y)<- vector
head(y)

#ABC order
class(y)
df_o<- y
###############
heat_plots(df_o, "Object")


###Season
physeq_genus
otu_table(physeq_genus)
#Calculate Relative Abundance
physeq.fam.rel <- microbiome::transform(physeq_genus, "compositional")

#convert to matrix 
y<- data.matrix(otu_table(physeq.fam.rel), rownames.force = NA)
head(y)
#Match with metadata
meta<- sample_data(physeq.fam.rel)
meta$sample= rownames(meta)
head(meta)

vector<- meta$Season[match(colnames(y), meta$sample)]
vector
colnames(y)<- vector
head(y)

#ABC order
class(y)
df_o<- y
###############
heat_plots(df_o, "Season")



###Area##########################################################
physeq_genus
otu_table(physeq_genus)
#Calculate Relative Abundance
physeq.fam.rel <- microbiome::transform(physeq_genus, "compositional")

#convert to matrix 
y<- data.matrix(otu_table(physeq.fam.rel), rownames.force = NA)
head(y)
#Match with metadata
meta<- sample_data(physeq.fam.rel)
meta$sample= rownames(meta)
head(meta)

vector<- meta$RoomType[match(colnames(y), meta$sample)]
vector
colnames(y)<- vector
head(y)

#ABC order
class(y)
df_o<- y
###############
heat_plots(df_o, "RoomType")





setwd("..")
getwd()

```


#Microbiome Network
```{r}
# convert to relative abundance
library(igraph)
setwd(("./Network_area")) 

######################Function#########################
network_taxa <- function(ps_object, name){
#Set seed for repdocucability
set.seed(99)
physeq_rel <- ps_object

#(Optional). Default 0.4. The maximum ecological distance (as defined by distance) allowed between two samples to still consider them “connected” by an edge in the graphical model
ig <- make_network(physeq_rel, dist.fun="bray", max.dist=0.8)
p2<- plot_network(ig, physeq_rel, color="RoomType", line_weight=1.5, label=NULL, point_size=12)+scale_colour_manual(values=color_factor)
write.csv (p2$data, paste(name,"Network Distance Bray-Curtis.csv"))


#Get info about network
print(neighborhood(ig))

#Diameter = lenght of longest connection 
diameter_ig<- diameter(ig)
#Number of nodes
numodes<- length(degree(ig))
#No edges
numdeg <- ecount(ig)

#Number of points connected to each vertex
degree(ig)
avgdeg <- numdeg / numodes

#Get centrality-The higher the network centralization, the more connected the network is
cent<- centralization.degree(ig)$centralization

#No. clusters
no_clusters<- clusters(ig)$no

vector= c(paste("diameter=", diameter_ig), paste("no. clusters", no_clusters), paste("centrality=", cent),paste("no nodes=", numodes), paste("no degrees=", numdeg), paste("Avg deg=", round(avgdeg,4)))
print(vector)

capture.output(vector , file = paste(name,"Network.txt"))


#Export plot to .tif file 
tiff(paste(name,"Bray-curtis Network.tif", sep=" "), res=300, compression = "lzw", height=9, width=9, units="in")
print(p2)
dev.off()

#Photo edit PDF in Inkscape 
pdf(paste(name,"Bray-curtis Network.pdf", sep=" "), height = 9, width = 9)
print(p2)
dev.off()

return(ig)
}
#########################################################
net_ig_NL<- network_taxa(ps_be_NL, "Built Environment_NL")
net_ig_B<- network_taxa(ps_be_B, "Built Environment_B")
net_ig_WS<- network_taxa(ps_be_WS, "Built Environment_WS")

```
