---
title: "Example ITS workflow all samples with good filtering"
author: "Sarah Daly"
date: '2024-02-14'
output:
  html_document: default
  word_document: default
editor_options:
  chunk_output_type: console
---

Input:
OTU or ASV table (.qza), unfiltered taxonomy table (.qza), rooted photogenic tree (.qza), metadata file (.txt)
Output
.csv and .tif 

All Sequencing Samples Summer Fall Winter Spring

Results focus on Fan/Walls/doors/floors 

#Set Working Directory
```{r}
setwd("C:/Users/dalys/Box/Microbial Food Safety & Spoilage Lab/Lab Members/Sarah Daly/ITS Methods Project/Github upload/Side_by_Side_Analysis")

list.files()
```

#Load Packages
There will be errors/warnings when you load packages but most you can ignore 
```{r}
library(scales)

library(DECIPHER) # This package will help in importing, maintaining, analyzing, manipulating, and exporting a massive amount of sequences.

library(ape) # Analyses of Phylogenetics and Evolution package. Required for tree calculations to be used with phyloseq

library(DESeq2) # This package will help analyze "differential expression" in the microbiota alongside phyloseq

library(ggplot2) # Graphing package used in phyloseq. To edit the default setting of a plot, you need to use functions in this package.

library(phyloseq) # The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.

library(plotly) # A package to create interactive web graphics of use in 3D plots

library(vegan) # The vegan package provides tools for descriptive community ecology. It has most basic functions of diversity analysis, community ordination and dissimilarity analysis. In general, this package is used for Bray-Curtis and Jaccard analyses.

library(tidyverse) # This package is designed to make it easy to install and load multiple 'tidyverse' packages in a single step

library(adespatial) # Tools for the multiscale spatial analysis of multivariate data

library(devtools) # Make package development easier by providing R functions that simplify and expedite common tasks

library(qiime2R) # A package for importing qiime artifacts into an R session

library(microbiome) # Data analysis and visualization

library(microbiomeSeq) # Data analysis and visualization

library("pander") # provide a minimal and easy tool for rendering R objects into Pandoc's markdown

library(ranacapa) # Data analysis 

library(grid) # support data visualization

library(gridExtra)  # support data visualization

library(knitr) # Provides a general-purpose tool for dynamic report generation in R using Literate Programming techniques.

library(png) # Figure download

library("ggdendro") #set of tools for dendrograms and tree plots using 'ggplot2'

library(ggpubr) # publication quality figures, based on ggplot2

library(RColorBrewer) # nice color options

library(microbiomeutilities) # some utility tools 

library(broom) #tidy anova tables 
library(hrbrthemes)
library(gcookbook)
library(tidyverse)
```

#Create a vector of colors for plotting
```{r}
nice_colors = c("#999999", "#E69F00", "#56B4E9","#e98756","firebrick3","#5800e6", "#CDDC49", "#C475D3", 
                "#E94B30", "#233F57", "#FEE659", "#A1CFDD", "#F4755E", "#D6F6F7","blueviolet", "#6898BF","cornsilk",  "tan3", "violetred4", "seashell3", "forestgreen", "grey60", "hotpink", "darkgreen", "dodgerblue3", "gold", "springgreen", "slateblue2", "darkolivegreen", "deeppink4", "gray16", "white", "yellow", "red", "tan", "slategray3", "wheat", "thistle4", "peachpuff", "plum", "orangered", "mediumseagreen", "hotpink4", "goldenrod3", "ivory", "darkturquoise", "lightgoldenrod", "violet", "firebrick", "mistyrose2", "brown", "rosybrown", "royalblue", "orange", "blue", "hotpink2", "tan4", "plum2", "grey40", "firebrick4", "violet", "brown4", "ivory2", "seashell4")
                
```

#Load Data and Convert to Phyloseq Object
```{r}
#Convert qiime artifacts directly to phyloseq
#phyloseq provides a set of classes and tools to facilitate the import, storage, analysis, and graphical display of microbiome census data.

#List files
list.files()

#Importing OTU/ASV table
ASVS <- read_qza("q20/otu-table-dn-99-q20.qza")   #Qiime output ASV file after filtering
dim(ASVS$data)

# Importing phylogenic tree
tree <- read_qza("q20/rooted_tree.qza")  #Qiime output file
tree$data

# Importing taxonomy assignment 
taxonomy <- read_qza("q20/unite-dyn-11-22-ITS-taxonomy-otu-q20.qza")  #Qiime output file

#Process taxonomy file more 
#get taxonomy names
tax_strings<- strsplit(as.character(taxonomy$data$Taxon), ";")
head(tax_strings)

#some strings are missing levels (blanks) and the dataframe will not bind
x = 1
y = 1

blanks <- c("k__","p__", "c__", "o__","f__", "g__", "s__")

num_taxa = length(tax_strings)  #number of taxa
num_taxa  #shoudl match dim OTUs$data

for (item in tax_strings){
  
  t = length(tax_strings[[x]]) #get list item length
  
if (t < 7)  {  #if less than 7 elements in each list item
  #print(t)
  y=1
  #print(taxa)
  while (y <= 7){  #loop through each item in item 
    if (is.na(tax_strings[[x]][y])){
      print('needs')
      tax_strings[[x]][y] = blanks[y]  #replace blank space with corr level character
      y=y+1
      }
      else(y=y+1)
    }
x = x+1
}
  else (x = x+ 1)
}

#convert to dataframe
tax_table<-as.data.frame(tax_strings)
dim(tax_table)
#Trasnpose table
tax_table<-t(tax_table)
head(tax_table)

#Add level labels
#Add oto/asv info to tax table
rownames(tax_table) <- taxonomy$data$Feature.ID
colnames(tax_table) <- c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
head(tax_table)
#Inspect table 
write.csv(tax_table, "merged_tax_table.csv")

#################Creating phyloseq object#######################################

#Note, phyloseq uses the term "OTU" loosely and since we fed it an ASV table, it is an ASV table
OTU = otu_table(as.matrix(ASVS$data), taxa_are_rows = T)
#Transpose OtU table
otu_table(OTU)<-t(otu_table(OTU))  #may need to do this again

TAX = tax_table(as.matrix(tax_table))
dim(TAX)
dim(taxa_names(OTU))
head(TAX)


# Importing metadata
metadata <- read.table("q20/Metadata_pairs.txt", sep='\t', header=T, row.names=1, comment="")
head(metadata)
dim(metadata)
length(sample_names(OTU))
SAMPLE = sample_data(metadata)

#phylogentic tree
TREE = tree$data

#Inspect files for troubleshooting 
write.csv(SAMPLE, "phyloseq SAMPLE input.csv")
write.csv(TAX, "phyloseq TAX input.csv")
write.csv(OTU, "phyloseq OTU input.csv")

# merge the data
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)
#contains OTU table, taxonomy table, sample data, and phy tree
#check that dimensions are consistent (same number taxa setwd("..")
#getwd()between OTU, tax_table)
ps
#######################Summary stats from phyloseq object########################
#Summarizing the phyloseq object to check for feature of data
# check for features of data  
summarize_phyloseq(ps)
print_ps(ps)
head(sample_data(ps))

```


#Filter Phyloseq Object
https://github.com/joey711/phyloseq/issues/1483 
```{r}
#####################################################
# Creating a table for number of features for each phyla
table(tax_table(ps)[, "Genus"], exclude = NULL)

# Removing the NAs and uncharacterized Phylums
ps0 <- subset_taxa(ps, !is.na(Genus) & !Genus %in% c("", "g__"))#Remove unidentified genera

# Summary Before and After
ps
ps0

# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(ps0),
               MARGIN = ifelse(taxa_are_rows(ps0), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps0),
                    tax_table(ps0))
# Viewing
head(prevdf)
# Computing average and total prevalence
temp_df <- plyr::ddply(prevdf, "Genus", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

newdata<- temp_df
colnames(newdata) <- c("Genus", "mean","sum_total")
head(newdata)

newdata<- newdata%>% arrange(desc(sum_total))
head(newdata)

write.csv(newdata, "Prevlance_Genus.csv")

# Vector to be removed
filterGenus <- temp_df[temp_df$`2` < 10,]$Genus  #Can change this number should be at least greater than 2

# Filter entries with unidentified Phylum.
ps1 = subset_taxa(ps0, !Genus %in% filterGenus)

# Summary
ps
ps0
ps1
```

#heat_map
```{r}
#Aggretate genus
physeq_genus <- aggregate_taxa(ps1, level = "Genus")

#Calculate Relative Abundance
physeq.fam.rel <- microbiome::transform(physeq_genus, "compositional")

#convert to matrix 
y<- data.matrix(otu_table(physeq.fam.rel), rownames.force = NA)
y
class(y)
head(y)
tail(y)

#Match with metadata
meta<- sample_data(physeq.fam.rel)
meta$sample= rownames(meta)
head(meta)

vector<- meta$Room_Site[match(colnames(y), meta$sample)]
vector
colnames(y)<- vector
head(y)

#ABC order
class(y)
df<- as.data.frame(y)
library(dplyr)

#ORder Columns
df2<- df[ , order(names(df))]
head(df2)

#Remove g__
row.names(df2) <- gsub("g__", "",rownames(df2))  # no libraries needed
head(df2)
df2 <- df2[order(row.names(df2)), , drop = FALSE]
head(df2)

#Then check each value< remove rows with no values > 5% RA
list.cols<- list()
keeps<- list()

i = 1
j = 1
while (i<= dim(df2)[1]){  #For each row
j = 1
list.cols<- list()
  while(j<= dim(df2)[2]){ #For each column
  list.cols[[j]] = df2[i,j] > 0.05
  j = j+1
  }
  
  #As long as one true keep
  vec<- unlist(list.cols)
  print(vec)
  print('new')
  if ('TRUE' %in% vec){keeps[[i]]= i}
  #else{keeps[[i]]== 'drop'}
    i = i+1
  }

print(keeps)
keeps<- unlist(keeps)
keeps  #Rows to remove

#Fitler to only have keeps
df3<- df2 %>% slice(keeps)
dim(df3)

y<- as.matrix(df3)

write.csv(df3, "heatmapdf.csv")

#Pivot data
head(df3)

df4 <- tibble::rownames_to_column(df3, "Genus")
head(df4)

#https://jcoliver.github.io/learn-r/006-heatmaps.html

mine_long <- pivot_longer(data = (df4), 
                          cols = -c("Genus"),
                          names_to = "Site", 
                          values_to = "Abundance")
head(mine_long)
tail(mine_long)

mine_long$Site_Room<- mine_long$Site
head(mine_long)

mine_long$Site_Room <- gsub('(.1*?).1.*', '\\1', mine_long$Site_Room)  # no libraries needed
head(mine_long)

mine_long$Abundance<- mine_long$Abundance*100
head(mine_long)

#PLot
mine_heatmap <- ggplot(data = mine_long, mapping = aes(x = Site,
                                                       y = Genus,
                                                       fill = Abundance)) +
  geom_tile(color = "gray") +
  #xlab(label = "Depth (m)") +
  facet_grid(~ Site_Room, switch = "x", scales = "free_x", space = "free_x") +
    scale_fill_gradient(name = "Relative Abundance",
                      low = "grey95",  ##FFFFFF
                      high = "darkblue", 
                      limits=c(0,80)) +  ##012345
  theme(strip.placement = "outside",
        plot.title = element_text(hjust = 0.5)) +
  #ggtitle(label = "Microbe Class Abundance")+
    theme(axis.text.y = element_text(color="black", size=16), axis.text.x = element_text(color="white", size=6))+
  theme(strip.text.x = element_text(size = 12, colour = "black"))


mine_heatmap

tiff("5 per Heatmap Genus.tif", res=300, compression = "lzw", height=8, width=9, units="in")
print(mine_heatmap)
dev.off()

pdf("5 per Heatmap Genus.pdf", height=8, width=9)
print(mine_heatmap)
dev.off()

write.csv(mine_long, "RA_heat.csv")


```

#Alpha Diveristy 
```{r}
setwd(("./Alpha_div"))

alpha_diversity_rare <- function(ps_object, name, group_var, alpha_vars){
  
set.seed(99)
print(ps_object)

#########################
alpha.div <- plot_richness(ps_object, x=group_var, measures=alpha_vars) +
    geom_boxplot() +
    theme_classic() +
    theme(text = element_text(size=24,colour="black",face="bold"),strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))


write.csv(alpha.div$data, paste(name, alpha_vars, group_var, "Dot Diversity.csv"))

}
###############################################################
#CALL Function 
#alpha_=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson")
#plot_ = c("diversity_inverse_simpson", "chao1", "observed","diversity_shannon","evenness_pielou", "evenness_simpson", "dominance_gini", "dominance_simpson", "diversity_gini_simpson")

alpha_diversity_rare(ps1, "Sidebyside", "Order_no", "Chao1" ) 
alpha_diversity_rare(ps1, "Sidebyside", "Order_no", "Shannon") 


setwd("..")
getwd()

```

